internal/jobcontainers/job.go

```diff
@@ -0,0 +1,308 @@
+		}
+		_, err := winapi.SetIoRateControlInformationJobObject(job.jobHandle, &ioInfo)
+		if err != nil {
+			return fmt.Errorf("failed to set IO limit info on job object: %s", err)
+		}
+	}
+	return nil
+}
+
+// createJobObject creates a job object, attaches an IO completion port to use
+// for notifications and then returns an object with the corresponding handles.
+func createJobObject(name string) (*jobObject, error) {
+	jobHandle, err := windows.CreateJobObject(nil, windows.StringToUTF16Ptr(name))
+	if err != nil {
+		return nil, err
+	}
+	iocpHandle, err := windows.CreateIoCompletionPort(windows.InvalidHandle, 0, 0, 1)
+	if err != nil {
+		windows.Close(jobHandle)
+		return nil, err
+	}
+	if _, err = attachIOCP(jobHandle, iocpHandle); err != nil {
+		windows.Close(jobHandle)
+		windows.Close(iocpHandle)
+		return nil, err
+	}
+	return &jobObject{jobHandle, iocpHandle}, nil
+}
+
+// close closes the job object and iocp handles. If this is the only open handle
+// the job object will be terminated.
+func (job *jobObject) close() error {
+	closeErr := false
+	if job.jobHandle != 0 {
+		if err := windows.Close(job.jobHandle); err != nil {
+			closeErr = true
+		}
+		job.jobHandle = 0
+	}
+	if job.iocpHandle != 0 {
+		if err := windows.Close(job.iocpHandle); err != nil {
+			closeErr = true
+		}
+		job.iocpHandle = 0
+	}
+	if closeErr {
+		return errors.New("failed to close one or more handles")
+	}
+	return nil
+}
+
+// Assign assigns a process to the job object.
+func (job *jobObject) assign(p *JobProcess) error {
+	if p.Pid() == 0 {
+		return errors.New("process has not started")
+	}
+	hProc, err := windows.OpenProcess(winapi.PROCESS_ALL_ACCESS, true, uint32(p.Pid()))
+	if err != nil {
+		return err
+	}
+	defer windows.Close(hProc)
+	return windows.AssignProcessToJobObject(job.jobHandle, hProc)
+}
+
+// Terminates the job, essentially calls TerminateProcess on every process in the
+// job.
+func (job *jobObject) terminate() error {
+	if job.jobHandle != 0 {
+		return windows.TerminateJobObject(job.jobHandle, 1)
+	}
+	return nil
+}
+
+func (job *jobObject) shutdown(ctx context.Context) error {
+	pids, err := job.pids()
+	if err != nil {
+		return fmt.Errorf("failed to get pids for job object: %s", err)
+	}
+	var (
+		terminate bool
+		signalErr bool
+	)
+	for _, pid := range pids {
+		if err := windows.GenerateConsoleCtrlEvent(windows.CTRL_BREAK_EVENT, pid); err != nil {
+			log.G(ctx).WithField("pid", pid).Error("failed to send ctrl-break to process in job")
+			signalErr = true
+		}
+	}
+
+	// Get pids in job again. If there is any left then terminate the job.
+	newPids, err := job.pids()
+	if err != nil {
+		return fmt.Errorf("failed to get pids for job object: %s", err)
+	}
+	terminate = len(newPids) != 0
+	// If any of the processes couldnt be killed gracefully just terminate the job.
+	// Equivalent to calling TerminateProcess on every proc in the job.
+	if terminate || signalErr {
+		return job.terminate()
+	}
```

Comment:
do we need to synchronize anything to prevent processes being added to the job during this cleanup?

Reply:
I synchronize in the container methods that aren't in this PR. Maybe it would be better to do the synch here

Reply:
hmm I guess it depends on how they're used. Do whatever makes the most sense to you I guess.

Reply:
I'd rather do the synch in the object that owns the handles I feel so I'll move it over to here